### PR TITLE
Issue12: Individual Memory

### PR DESCRIPTION
**Pull request by shamith2**
#12

### CODE REVIEW

File: gym_cap/envs/agent.py [hunk 1]
```diff
@@ -28,6 +30,7 @@ def __init__(self, loc, map_only, team_number):
         self.range = UGV_RANGE
         self.a_range = UGV_A_RANGE
         self.air = False
+        self.memory = np.full_like(self._env, -1)
```
Comment (skim0119) [inline gym_cap/envs/agent.py]: I don't think this would run.

File: gym_cap/envs/agent.py [hunk 2]
```diff
@@ -87,6 +90,28 @@ def move(self, action, env, team_home):
...
 
+    def update_memory(self, obs):
+
+        """
+        saves/updates individual map of an agent
+
+        """
+        
+        leng, breth = obs.shape
+        l, b = floor(leng/2), floor(breth/2)
```
Comment (skim0119) [inline gym_cap/envs/agent.py]: Just use ```leng//2``` instead of using math package

File: gym_cap/envs/cap_env.py [hunk 3]
```diff
@@ -385,7 +385,7 @@ def seed(self, seed=None):
         """
         self.np_random, seed = seeding.np_random(seed)
         return [seed]
-
+
```
Comment (skim0119) [inline gym_cap/envs/cap_env.py]: Try to remove commits that is not part of the change

**Review by skim0119 — CHANGES_REQUESTED** [review]
On configuration setting, add a parameter that sets different mode of individual memory.

1. Fog : it builds the memory based on the observation. (What you currently made)
2. None : No memory. All memory should be -1 (unknown)
3. Full : Memory should be same as static map.

On environment step() method, add 'update memory' for individual agent if the above parameter is set to 'Fog.'

Currently, the code is not running.

File: gym_cap/envs/agent.py (hunk 1 above)
Comment (shamith2) [reply gym_cap/envs/agent.py]: changed to self.memory = np.array([[-1 for i in range(self.map_size[0])] for j in range(self.map_size[1])])

File: gym_cap/envs/agent.py (hunk 2 above)
Comment (shamith2) [reply gym_cap/envs/agent.py]: changed accordingly

File: gym_cap/envs/agent.py [hunk 4]
```diff
@@ -24,10 +25,13 @@ def __init__(self, loc, map_only, team_number):
         """
         self.isAlive = True
         self.x, self.y = loc
+        self.length, self.breath = map_size
```
Comment (skim0119) [inline gym_cap/envs/agent.py]: map_size --> map_only.shape

File: gym_cap/envs/cap_env.py [hunk 5]
```diff
@@ -493,6 +493,10 @@ def step(self, entities_action=None, cur_suggestions=None):
         self.state = np.copy(self._env)
 
         isDone = self.red_win or self.blue_win
+        
+        if Agent.mode == "fog":
+            Agent.update_memory(obs)
```
Comment (skim0119) [inline gym_cap/envs/cap_env.py]: For all individual red and blue

**Review by skim0119 — CHANGES_REQUESTED** [review]

File: gym_cap/envs/agent.py (hunk 1 above)
Comment (skim0119) [reply gym_cap/envs/agent.py]: use map_only.shape

File: gym_cap/envs/cap_env.py (hunk 3 above)
Comment (shamith2) [reply gym_cap/envs/cap_env.py]: removed commit

File: gym_cap/envs/agent.py (hunk 4 above)
Comment (shamith2) [reply gym_cap/envs/agent.py]: changed accordingly

File: gym_cap/envs/agent.py (hunk 1 above)
Comment (shamith2) [reply gym_cap/envs/agent.py]: changed accordingly

File: gym_cap/envs/cap_env.py (hunk 5 above)
Comment (skim0119) [reply gym_cap/envs/cap_env.py]:
``` bash
Traceback (most recent call last):
  File "cap_test.py", line 39, in <module>
    observation, reward, done, info = env.step()  # feedback from environment
  File "/Users/skim0119/github/raide_ctf/gym_cap/envs/cap_env.py", line 497, in step
    if Agent.mode == "fog":
AttributeError: type object 'Agent' has no attribute 'mode'
```

Comment (shamith2) [reply gym_cap/envs/cap_env.py]: added 'mode' attribute

Comment (shamith2) [reply gym_cap/envs/cap_env.py]: added 'mode' attribute

File: gym_cap/envs/agent.py [hunk 6]
```diff
@@ -25,12 +25,12 @@ def __init__(self, loc, map_only, team_number):
...
         self.x, self.y = loc
-        self.length, self.breath = map_size
+        self.length, self.breath = map_only.shape
         self.step = UGV_STEP
         self.range = UGV_RANGE
         self.a_range = UGV_A_RANGE
         self.air = False
-        self.memory = np.array([[-1 for i in range(self.length)] for j in range(self.breath)])
+        self.memory = np.full((self.length, self.breath), -1)
         self.mode = "fog"
```
Comment (skim0119) [inline gym_cap/envs/agent.py]: Set default to None

File: gym_cap/envs/agent.py [hunk 7]
```diff
@@ -91,6 +91,9 @@ def move(self, action, env, team_home):
         else:
             print("error: wrong action selected")
 
+    def mode(self):
```
Comment (skim0119) [inline gym_cap/envs/agent.py]: Please remove

**Review by skim0119 — CHANGES_REQUESTED** [review]

File: gym_cap/envs/agent.py (hunk 7 above)
Comment (shamith2) [reply gym_cap/envs/agent.py]: removed mode

File: gym_cap/envs/agent.py (hunk 6 above)
Comment (shamith2) [reply gym_cap/envs/agent.py]: set default value of memory_mode to None

File: gym_cap/envs/agent.py [hunk 8]
```diff
@@ -86,7 +90,30 @@ def move(self, action, env, team_home):
...
             print("error: wrong action selected")
+    
+    def update_memory(self):
+
+        """
+        saves/updates individual map of an agent
 
+        """
+        
+        obs = self.get_obs()
```
Comment (skim0119) [inline gym_cap/envs/agent.py]: Use the following:
```py
get_obs(self, com_ground, com_air, env, distance, freq, *args)
```

All parameters will have default values except env.
So, the code should be simply ```self.get_obs(env=env)```

We need 'env,' so add env parameter for update_memory method.

File: gym_cap/envs/cap_env.py [hunk 9]
```diff
@@ -493,6 +493,11 @@ def step(self, entities_action=None, cur_suggestions=None):
         self.state = np.copy(self._env)
 
         isDone = self.red_win or self.blue_win
+        
+        for agent in self.team_blue + self.team_red:
+            if agent.memory_mode == "fog":
+                Agent.update_memory()
```
Comment (skim0119) [inline gym_cap/envs/cap_env.py]: Agent.update_memory causes an issue.
update_memory needs to be called for individual agents.

Also, please pass the environment parameter.
```py
agent.update_memory(env=self)
```

**Review by skim0119 — CHANGES_REQUESTED** [review]

File: gym_cap/envs/agent.py (hunk 8 above)
Comment (shamith2) [reply gym_cap/envs/agent.py]: added 'env' parameter to update_memory

**Review by skim0119 — COMMENTED** [review]
get_obs method is now placed in agent.

Please double check your methods and see if there is any issues.

**Review by skim0119 — APPROVED** [review]
